src/main/scala/tech/cryptonomic/conseil/tezos/MetadataCaching.scala

```diff
@@ -0,0 +1,137 @@
+package tech.cryptonomic.conseil.tezos
+
+import cats.Monad
+import cats.effect.Concurrent
+import cats.effect.concurrent.{MVar, Ref}
+import com.rklaehn.radixtree.RadixTree
+import tech.cryptonomic.conseil.generic.chain.PlatformDiscoveryTypes.{Attribute, Entity}
+import tech.cryptonomic.conseil.tezos.MetadataCaching._
+
+/** Companion object providing useful types related with MetadataCaching */
+object MetadataCaching {
+  type LastUpdated = Long
+
+  type MetadataCache[A] = Map[String, (LastUpdated, A)]
+  type AttributesCache = MetadataCache[List[Attribute]]
+  type AttributeValuesCache = MetadataCache[RadixTree[String, String]]
+  type EntitiesCache = MetadataCache[List[Entity]]
+
+  /** Cache initialization statuses */
+  sealed trait CachingStatus extends Product with Serializable
+
+  case object NotStarted extends CachingStatus
+
+  case object InProgress extends CachingStatus
+
+  case object Finished extends CachingStatus
+  import cats.implicits._
+
+  /** Initializes metadata caching */
+  def empty[F[_]](implicit concurrent: Concurrent[F]): F[MetadataCaching[F]] =
+    for {
+    cachingStatus <- Ref[F].of[CachingStatus](NotStarted)
+    attributesCache <- MVar[F].empty[AttributesCache]
+    entitiesCache <- MVar[F].empty[EntitiesCache]
+    attributeValuesCache <- MVar[F].empty[AttributeValuesCache]
+  } yield new MetadataCaching[F](cachingStatus, attributesCache, entitiesCache, attributeValuesCache)
+
+  /** Returns current time in milliseconds */
+  def now: Long = System.nanoTime()
+
+  /** Makes key out of table and column names */
+  def makeKey(table: String, column: String): String = s"$table.$column"
+}
+
+/** Class providing caching for metadata */
+class MetadataCaching[F[_]](
```

Comment:
Have you checked [ScalaCache](https://github.com/cb372/scalacache)? It supports [Caffeine](https://github.com/blemale/scaffeine) which has support for [cache refreshing](https://github.com/ben-manes/caffeine/wiki/Refresh).

Reply:
At the point when I started implementing cache by myself the logic required custom solution.
I'll look into it and check if ScalaCache can help now.